stdlib/asm/crypto/stark/deep_queries.masm

```diff
@@ -3,782 +3,388 @@ const.MAIN_TRACE_COM_PTR=4294967290
     ext2mul
-    #=> [prod1, prod0, - T_i(gz)_1, value_i - T_i(gz)_0, beta1, beta0, prod1_g, prod0_g, Acc, value_i, p1, p2, p3]
-    # where (prod0, prod1) =  (value_i - T_i(z)_0, - T_i(z)_1) * beta = (T_i(x) - T_i(z)) * beta
-    movup.7
-    movup.7
-    dupw.2
-    #=> [Acc, prod1_g, prod0_g, prod1, prod0, y, y, y, y, Acc, value_i, p1, p2, p3]
-
-    # Accumulate
-    movup.7
-    movup.7
-    ext2add
-    movdn.5
-    movdn.5
-    ext2add
-    swap
-    swap.3
-    swap
-    swap.2
-    #=> [Acc3 + prod1, Acc2 + prod0, Acc1 + prod1_g, Acc0 + prod0_g, y, y, y, y, Acc, value_i, p1, p2, p3]           Acc = [Tz, Tz, Tgz, Tgz]
-    #=> [Acc, Y, Y, P]
-
-    # Prepare for next cycle, if any
-    swapw.3
-    swapw
-    #=> [Y, P, Y, Acc]
+    #=> [j1', j0', i1', i0', CURRENT_TRACE_ROW_PTR, OOD_TRACE_PTR + 1, DEEP_RAND_CC_PTR + 2, 0, ...]
 end
 
-
-
-#! Compute the deep composition polynomial at the previously drawn random queries. Currently, their
-#! number is fixed to 27.
+#! Compute the deep composition polynomial at the previously drawn random queries.
 #!
-#! Input: [...]
+#! Input: [num_queries, ...]
 #! Output: [...]
-#! Cycles: 44 + 27 * 8445 = 228059
+#! Cycles: 17 + num_queries * 590
 export.compute_deep_composition_polynomial_queries
     push.FRI_QUERIES_PTR
-    repeat.5 push.0.0.0.0 end
+    add
     push.FRI_QUERIES_PTR
-    #=>[ptr_query, Y, Y, Y, Y, Y, ptr_query]
+    swap
+    push.0.0.0.0
+    push.FRI_QUERIES_PTR
+    #=>[ptr_query, Y, ptr_query_end, ptr_query]
 
-    repeat.27
+    push.1
+    while.true
         # Main trace portion of the query
+
         ## Get the next query index
         mem_loadw
         #=> [query, depth, y, y, ..]
+
         ## Get main trace commitment and use it to get the leaf
-        swapw
+        push.0.0.0.0
         mem_loadw.MAIN_TRACE_COM_PTR
-
+        #=>[R, query, depth, y, y, ptr_query_end, ptr_query]
         swapw
         mem_storew.TMP4
         movup.2 drop
         movup.2 drop
         swap
-        #=> [d, query, Com, Y, Y]
+        #=>[depth, query, R, ptr_query_end, ptr_query]
+        
         ## Get the leaf in the main trace commitment and save it
         mtree_get
         mem_storew.TMP3
         adv.keyval
-        #=> [V, R, Y, Y]
+        #=>[V, R, ptr_query_end, ptr_query]
         drop
         push.CURRENT_TRACE_ROW_PTR
-        swapw.3
-        #=> [Y, Y, Y, ptr, y, y, y ]
-        mem_loadw.ZERO_WORD_PTR
-        swapw.2
-        #=> [Y, Y, 0, 0, 0, 0, ptr, y, y, y]
+        swapw
+        #=>[R, ptr, y, y, y, ptr_query_end, ptr_query]
+        mem_loadw.ZERO_ZERO_ZERO_ONE_PTR
+        push.0.0.0.0
+        push.0.0.0.0
+        #=> [Y, Y, 0, 0, 0, 1, ptr, y, y, y]
         repeat.9
             adv_pipe hperm
         end
-        #=> [Y, L, Y, ptr, y, y, y]
+        dropw
+        adv_push.1
+        push.1
+        push.0.0
+        dup.12 mem_storew
+        swapw
+        mem_loadw.ZERO_WORD_PTR
```

Comment:
Would it make sense to move these (and a few following instructions) into a helper procedure? Maybe something like:
```
#! Input: [query_idx, ...]
#! Output: [...]
proc.load_main_query
```
We could do something similar for the auxiliary trace queries as well.

Reply:
Yes we could but it would incur at least 8 * 3 * 27 = 648 additional cycles for 27 queries.

Reply:
So, basically 24 extra cycles per query? What would be cause of this? (or maybe there is a better way to define a procedure or two which would encapsulate some of the logic from here without extra cost?)

Reply:
Yes - let's do that!